caikit/core/model_management/model_trainer_base.py

```diff
@@ -75,14 +114,52 @@ def wait(self):
         def load(self) -> ModuleBase:
             """A model future must be loadable with no additional arguments"""
 
+        ## Common Impl ##
+
+        @classmethod
+        def _save_path_with_id(
+            cls,
+            save_path: Optional[str],
+            save_with_id: bool,
+            training_id: str,
+        ) -> Optional[str]:
+            """If asked to save_with_id, child classes should use this shared
+            utility to construct the final save path
+            """
+            if save_path is None:
+                return save_path
+            if not save_with_id or training_id in save_path:
+                return save_path
+
+            # If told to save with the ID in the path, inject it right before the
+            # final portion of the path which is assumed to be the model ID.
+            path_parts = os.path.split(save_path)
+            return os.path.join(
+                *list(path_parts[:-1] + (training_id,) + path_parts[-1:])
+            )
+
     @abc.abstractmethod
     def train(
         self,
         module_class: Type[ModuleBase],
         *args,
         save_path: Optional[str] = None,
+        save_with_id: bool = False,
```

Comment:
Currently we were not using / providing `save_path` to `.train` function when training and essentially considering `save` a totally different and explicit operations. However, this has a limitation that we can't have a way to save or do something with iterations of the model checkpoints that are getting generated at train time. This could be useful for couple of purposes like when training is taking longer time, so we can get subsequent and iterative checkpoints,  these can also be used in some trustworthy AI pipelines, could be a fail safe mechanism, can be used to allow evaluation at multiple epochs at a later stage. 

Since now we are adding `save_path` and `save_with_id` both in overall `.train` function of model-management, do you think this will facilitate the usecases I mentioned above, essentially that allows persisting artifacts at the time of `.train`  function itself?